pad_server/src/controller.c

```diff
@@ -102,72 +101,88 @@ static ssize_t controller_recv(controller_t *controller, void *buf, size_t n) {
+            size_t total_read = 0;
+
+            while (total_read < sizeof(hdr)) {
+                bread = controller_recv(&controller, (char *)&hdr + total_read, sizeof(hdr) - total_read);
+                if (bread == -1) {
+                    if (errno == EINTR) continue; // Interrupted, try again
+                    fprintf(stderr, "Error reading message header: %s\n", strerror(errno));
+                    break;
+                } else if (bread == 0) {
+                    fprintf(stderr, "Control box disconnected.\n");
+                    break;
+                }
+                total_read += bread;
+            }
 
-        switch ((packet_type_e)hdr.type) {
-        case TYPE_CNTRL:
+            if (bread <= 0) break; // something that should not have happened happened, try reconnecting just in case
+
+            // TODO: handle recv errors
+
+            switch ((packet_type_e)hdr.type) {
+            case TYPE_CNTRL:
+
+                switch ((cntrl_subtype_e)hdr.subtype) {
+                case CNTRL_ACT_ACK:
+                case CNTRL_ARM_ACK:
+                    fprintf(stderr, "Unexpectedly received acknowledgement from sender.\n");
+                    thread_return(EXIT_FAILURE);
+                    break;
+                case CNTRL_ACT_REQ: {
+                    act_req_p req;
+                    controller_recv(&controller, &req, sizeof(req));
+                    printf("Received actuator request for ID #%u and state %s.\n", req.id, req.state ? "on" : "off");
+                } break;
+                case CNTRL_ARM_REQ: {
+                    arm_req_p req;
+                    controller_recv(&controller, &req, sizeof(req));
+                    printf("Received arming state %u.\n", req.level);
+                } break;
+                }
 
-            switch ((cntrl_subtype_e)hdr.subtype) {
-            case CNTRL_ACT_ACK:
-            case CNTRL_ARM_ACK:
-                fprintf(stderr, "Unexpectedly received acknowledgement from sender.\n");
+                break;
+            default:
+                fprintf(stderr, "Invalid message type: %u\n", hdr.type);
                 thread_return(EXIT_FAILURE);
                 break;
-            case CNTRL_ACT_REQ: {
-                act_req_p req;
-                controller_recv(&controller, &req, sizeof(req));
-                printf("Received actuator request for ID #%u and state %s.\n", req.id, req.state ? "on" : "off");
-            } break;
-            case CNTRL_ARM_REQ: {
-                arm_req_p req;
-                controller_recv(&controller, &req, sizeof(req));
-                printf("Received arming state %u.\n", req.level);
-            } break;
             }
-
-            break;
-        default:
-            fprintf(stderr, "Invalid message type: %u\n", hdr.type);
-            thread_return(EXIT_FAILURE);
-            break;
         }
+
+        // Disconnect before attempting to reconnect
+        controller_disconnect(&controller);
```

Comment:
Have you tried without this call? If we get here it's because the controller disconnected for some reason; shouldn't we be able to just re-connect?